--- a/package/firmware/linux-firmware/Makefile
+++ b/package/firmware/linux-firmware/Makefile
@@ -15,7 +15,7 @@
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_DATE:=2026-03-12
 PKG_SOURCE_VERSION:=06cd1ccba74a0e82edcc81506b9d6c7aac558849
-PKG_MIRROR_HASH:=c2a1c43b1ec530c2df0acb000b4dd7d70d9a5b29b1e20e88c44db1bb81101620
+PKG_MIRROR_HASH:=da1c492cf90fb909fa264efacb80b55f267fc8388343a5cb35b41936ea375532
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 
