--- a/package/firmware/linux-firmware/Makefile
+++ b/package/firmware/linux-firmware/Makefile
@@ -15,7 +15,7 @@
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_DATE:=2026-02-04
 PKG_SOURCE_VERSION:=2ebbf1b16029b087dfadf1bd2049145130560e14
-PKG_MIRROR_HASH:=b8c1c720ab5397a73e1e856669bfb29a8ee932c0e97b5071575fe0ca46491457
+PKG_MIRROR_HASH:=e3a8455d595807cf777103b760c8a85bbde06ec811ce7d12ea04d7c0f69fc2e6
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 
